feat: Add support for addons

tnimni opened the pull request:
# PR o'clock

## Description

Add support for eks addons recommended by was
1. VPC CNI
2. CoreDns
3. kube-proxy

The default behavior is to install the addons to the cluster.

Requires irsa to be enabled to deploy the VPC CNI addon

### Checklist

- [V] README.md has been updated after any changes to variables and outputs. See https://github.com/terraform-aws-modules/terraform-aws-eks/#doc-generation

daroga0002 commented on modules/addons/variables.tf:
```diff
@@ -0,0 +1,83 @@
+# Hack for a homemade `depends_on` https://discuss.hashicorp.com/t/tips-howto-implement-module-depends-on-emulation/2305/2
+# Will be removed in Terraform 0.13 with the support of module's `depends_on` https://github.com/hashicorp/terraform/issues/10462
+variable "eks_depends_on" {
+  description = "List of references to other resources this submodule depends on."
```
maybe this should me moved to module level as terraform from some time allowing it?

tnimni replied:
This pattern is used in modules/fargate and modules/node_groups, I just copied it from there to keep everything consistent.
I'm not sure which example you require, as eks_depends_on is defined in addons.tf in the root of the main module, https://github.com/tnimni/terraform-aws-eks/blob/feature/addons/addons.tf#L11-L14

daroga0002 reviewed (commented):
can we add some example into `examples` directory which will be using this?

daroga0002 commented on modules/addons/variables.tf:
```diff
@@ -0,0 +1,83 @@
+
+variable "coredns_versions" {
+  # Versions are taken from https://docs.aws.amazon.com/eks/latest/userguide/managing-coredns.html#updating-coredns-add-on
+  type        = map(any)
```
I think better will be `map(string)` just to not loose type over time

tnimni replied:
Changed to `map(string)`

daroga0002 commented on modules/addons/variables.tf:
```diff
@@ -0,0 +1,83 @@
+
+variable "kube_proxy_versions" {
+  # Versions are taken from https://docs.aws.amazon.com/eks/latest/userguide/managing-kube-proxy.html#updating-kube-proxy-add-on
+  type        = map(any)
```
I think better will be `map(string)` just to not loose type over time

tnimni replied:
Changed to `map(string)`

daroga0002 commented on modules/addons/variables.tf:
```diff
@@ -0,0 +1,83 @@
+variable "vpc_cni_versions" {
+  # Versions are taken from https://docs.aws.amazon.com/eks/latest/userguide/managing-vpc-cni.html#updating-vpc-cni-add-on
+  # Latest patch version is taken from https://github.com/aws/amazon-vpc-cni-k8s
+  type        = map(any)
```
I think better will be `map(string)` just to not loose type over time

tnimni replied:
Changed to `map(string)`

daroga0002 commented on modules/addons/variables.tf:
```diff
@@ -0,0 +1,83 @@
+# Will be removed in Terraform 0.13 with the support of module's `depends_on` https://github.com/hashicorp/terraform/issues/10462
+variable "eks_depends_on" {
+  description = "List of references to other resources this submodule depends on."
+  type        = any
```
probably `string` will be better

tnimni replied:
based on https://github.com/tnimni/terraform-aws-eks/blob/feature/addons/modules/fargate/variables.tf#L59-L65 , I copied it to keep everything consistent

daroga0002 replied:
yup, this was created before modules were supporting dependncies. Lets maintainers decide does we want to leave this for consistency or we will start using module dependency.

daroga0002 commented on modules/addons/outputs.tf:
```diff
@@ -0,0 +1,14 @@
+output "vpc_cni_id" {
+  description = "The id of the Amazon VPC CNI addon"
+  value       = aws_eks_addon.vpc_cni[0].id
```
does it will be working when I will not create this addon (there will be no element 0)?

tnimni replied:
Fixed the count=0 for all addons

daroga0002 commented on modules/addons/coredns.tf:
```diff
@@ -0,0 +1,12 @@
+  cluster_name      = var.cluster_name
+  addon_name        = "coredns"
+  resolve_conflicts = "OVERWRITE"
+  addon_version     = lookup(var.coredns_versions, var.cluster_version, "Not Found")
```
how AWS API will react when we pass `"Not Found"`, does there is validation on their end?

tnimni replied:
This is on purpose, it will not even reach aws api. terraform plan will throw an error stating the addon version can not be `"Not Found"` giving a hint for the user that the cluster version is not supported by the module

daroga0002 reviewed (commented):
for me it looks good, @barryib  can you take a look on it?

stevehipwell commented:
Please can the defaults not be to enable these.

tnimni commented:
@stevehipwell 
changed defaults to false

tnimni commented:
Get we llease get this approved? @barryib

thecosmicfrog commented:
Good work here @tnimni - hoping for it to be merged soon.

tnimni commented:
> Good work here @tnimni - hoping for it to be merged soon.

Thanknyo

> Good work here @tnimni - hoping for it to be merged soon.

Thank you

mmerickel commented:
So if we enable the vpc cni addon then should we pass `attach_worker_cni_policy=false`? Would be good to mention that in the docs if so.

tnimni commented:
> attach_worker_cni_policy=false

Hi,
it is not required to do so,

I checked and aws recommend it here https://docs.aws.amazon.com/eks/latest/userguide/cni-iam-role.html

I added to the example and also updated the main module readme to reflect that the variable attach_worker_cni_policy should be set to false if deploying the vpc cni addon

I appreciate your input

nikitacr7 commented:
Hi

Awesome. We're waiting for it.

thapakazi commented:
Same Same, nice work @tnimni  :clap:

tnimni commented:
@nikitacr7 @thapakazi thank you both

csschwe commented:
@tnimni This is great!!!

I do have one request though.  Can you add support for tags like the node_groups module

https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/node_groups.tf#L9

Thanks

tnimni commented:
> @tnimni This is great!!!
> 
> I do have one request though. Can you add support for tags like the node_groups module
> 
> https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/node_groups.tf#L9
> 
> Thanks

I'll have a look into it

wellermann commented:
Agreed, this is great.

EKS 1.21 just came out this week so it might be worth adding it now.

Either way this is great work @tnimni Thank you for doing this.

aftouh commented:
Awesome. I'm waiting for it.
Thank you @tnimni.

kuntalkumarbasu commented:
You just save a ton of my time. thank you @tnimni .
When can we get it merged?

tnimni commented:
> You just save a ton of my time. thank you @tnimni .
> When can we get it merged?

Hi,
I'm not able to merge or approve
need to ask @barryib

tnimni commented:
> EKS 1.21 just came out this week so it might be worth adding it now.
> 
> Either way this is great work @tnimni Thank you for doing this.

Hi,
I added it, also updated older cluster versions with the latest recommended addons

stevehipwell commented:
@tnimni I've just reviewed this and the whole addons module should be conditional, not the resources inside it.

stevehipwell commented:
@tnimni my second observation is that this would introduce a significant maintenance overhead to the module for it to function as expected without user interaction. Every time a new addon version is released you'd need to update the module addon version maps or else the paved road for users would be to get old versions. You'd also need to change the module for each new EKS version. The code for this module hasn't been modified for about 2 months at this point so I really don't think that the current design works.

If the user is providing the addon versions they have to understand the addon release cycle by which point they would be better off coding it themselves; as is this pattern adds more overhead as you need to know the state of the module as well as the state of the addon releases.

There is the [aws eks describe-addon-versions](https://docs.aws.amazon.com/cli/latest/reference/eks/describe-addon-versions.html) command to get the available addon versions for a specific EKS version, meaning that the Terraform provider should be able to support this as a data source. I'd suggest holding off on this PR until this is available as it'd solve the above concerns.

bryantbiggs commented:
hi all 👋🏼 - this looks interesting, but also a bit scary. the EKS module is already quite large and complex, we've even discussed how best to move forward since it is becoming so large and complex. I don't know if adding this functionality onto the module directly is the best course of action - this seems to be more of an independent module or a sub-module that is standalone and not tied into the root of the overall EKS module.

@terraform-aws-modules/triage-supporters what are your thoughts?
- add as independent sub-module
- do not add to this module
- other?

antonbabenko commented:
Good point, Bryant! We should try not to increase the complexity by adding more and more features if they don't belong there but I think EKS-addons _belong_ to the root module.

I think this PR can be simplified by having one resource `aws_eks_addon` in the root module which will accept a map of elements (one per addon). There is no need to maintain it as a separate submodule.

nitrocode commented:
@antonbabenko like this?

```hcl
module "eks" {
  # ...
  addons = [
    {
      addon_name               = "vpc-cni"
      addon_version            = "v1.8.0-eksbuild.1"
      resolve_conflicts        = "NONE"
      service_account_role_arn = null
    }
  ]
  # ...
}
```

antonbabenko commented:
@nitrocode Yes, but instead of a `list` it can be `map` (benefits: no need to make a conversion in `for_each` and it indicates that key in the map will be used as a key for resource):

```hcl
module "eks" {
  # ...
  addons = {
    "vpc-cni" = {
      addon_version     = "v1.8.0-eksbuild.1"
      resolve_conflicts = "NONE"
    }
  }
  # ...
}
```

stevehipwell commented:
Thanks @antonbabenko your design deals with my concerns and will allow anyone who wants to use addons to do so without impacting any of us who choose not to.

daroga0002 commented:
adding https://github.com/hashicorp/terraform-provider-aws/issues/20725 as if this will be added to provider will allow to remove `coredns_versions`, `kube_proxy_versions` and `vpc_cni_versions` variables

ivankovnatsky commented:
still needed.

bryantbiggs commented:
not stale

beingamarnath commented:
Any updates on this? Now that EKS version 1.21 is out there, can we introduce this to the module directly

Lillecarl commented:
Not stale

antonbabenko commented:
This issue has been resolved in version 18.0.0 :tada:

mmerickel commented:
I think this was closed in error... I do not see any of the changes from this PR in master or the 18.0.3 release.

bryantbiggs commented:
@mmerickel https://github.com/terraform-aws-modules/terraform-aws-eks/blob/3c6686379de9489d4c044cd13a63e2329e2ae4e1/main.tf#L220

bryantbiggs commented:
yes, see convo above https://github.com/terraform-aws-modules/terraform-aws-eks/pull/1443#issuecomment-897416141

there are examples of using this features as well https://github.com/terraform-aws-modules/terraform-aws-eks/blob/3c6686379de9489d4c044cd13a63e2329e2ae4e1/examples/complete/main.tf#L29
